### PR TITLE
Add match score functions

### PR DESCRIPTION
**Pull request by grthomson**
## Description
Add match score functions and their unit tests from old repository.

 
## Type of change

*You can delete options that are not relevant.*

- [ ] Bug fix - *non-breaking change*
- [x] New feature - *non-breaking change*
- [ ] Breaking change - *backwards incompatible change, changes expected behaviour*
- [ ] Non-user facing change, structural change, dev functionality, docs ...

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented my code appropriately, focusing on explaining my design decisions (explain why, not how).
- [x] I have made corresponding changes to the documentation (comments, docstring, etc.. )
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have updated the change log.

<br>

#  Peer review
Any new code includes all the following:

- **Documentation**: docstrings, comments have been added/ updated.
- **Style guidelines**: New code conforms to the project's contribution guidelines.
- **Functionality**: The code works as expected, handles expected edge cases and exceptions are handled appropriately.
- **Complexity**: The code is not overly complex, logic has been split into appropriately sized functions, etc..
- **Test coverage**: Unit tests cover essential functions for a reasonable range of inputs and conditions. Added and existing tests pass on my machine.

### Review comments
Suggestions should be tailored to the code that you are reviewing. Provide context.
Be critical and clear, but not mean. Ask questions and set actions.
<details><summary>These might include:</summary>

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented
  - Do the tests effectively assure that it
  works correctly? Are there additional edge cases/ negative tests to be considered?
- code style improvements (could the code be written more clearly?)
</details>
<br>

*Further reading: [code review best practices](https://best-practice-and-impact.github.io/qa-of-code-guidance/peer_review.html)*

### CODE REVIEW

File: scalelink/match_scores/match_scores.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+            variables.
+        x_star_scaled (dict of str: float):
+            A dictionary with keys consisting of linkage variable names and
+            states and the values consisting of the scaled weights calculated for
+            those variables and states.
+        spark (PySpark SparkSession):
+            Name of the Spark session being used.
+
+    Dependencies:
+        re
```
Comment (mary-cleaton) [inline scalelink/match_scores/match_scores.py]: As the dependency on `re` is not present, this should be removed.

File: scalelink/match_scores/match_scores.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+            cut-points. Dictionary keys are the names of the input variables and
+            dictionary values are the values of the variables themselves.
+            This may be produced by the utils function get_input_variables().
+        spark (PySpark SparkSession):
+            Name of the Spark session being used.
+
+    Dependencies:
+        re
+        match_scores.assign_weights
+        match_scores.assign_match_score
```
Comment (mary-cleaton) [inline scalelink/match_scores/match_scores.py]: These are no longer true.

File: tests/match_scores/test_match_scores.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+  test_assign_weights
+
+  test_get_match_scores - test shell only, currently
+"""
+
+import chispa as ch
+import pytest
+
+from scalelink.match_scores import match_scores as msf
```
Comment (mary-cleaton) [inline tests/match_scores/test_match_scores.py]: Following the example of the other function and test scripts, this should be aliased as `ms`.

File: tests/match_scores/test_match_scores.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+
+def test_assign_match_score(spark):
+    """
+    Tests that assign_match_score() gives the correct output when provided with
+    suitable inputs.
```
Comment (mary-cleaton) [inline tests/match_scores/test_match_scores.py]: Fussy nonsense request, but please can you make sure that all test function docstrings use the following format:

`"""Tests that <function> gives the correct output when provided with appropriate inputs."""`

**Review by mary-cleaton — CHANGES_REQUESTED** [review]
Overall, looks good. Everything that should be there is there and there are only the expected minor changes. A few minor corrections requested.

File: scalelink/match_scores/match_scores.py (hunk 2 above)
Comment (grthomson) [reply scalelink/match_scores/match_scores.py]: Removed! I guess the latter two aren't actually "dependencies" in the sense of external requirements. Also the docstring at the top mentions that they are run by get_match_scores.

File: scalelink/match_scores/match_scores.py (hunk 1 above)
Comment (grthomson) [reply scalelink/match_scores/match_scores.py]: Removed.

File: scalelink/match_scores/match_scores.py (hunk 2 above)
Comment (mary-cleaton) [reply scalelink/match_scores/match_scores.py]: Yeah, they originally kinda were when those functions were in a different script. Not really any more now that they're all in the same script.

**Review by mary-cleaton — APPROVED** [review]
Happy that all requested changes have been made.
